truffle/src/main/java/org/jruby/truffle/nodes/core/RegexpNodes.java

```diff
@@ -335,31 +337,42 @@ public static Encoding checkEncoding(DynamicObject regexp, CodeRangeable str, bo
         return enc;
     }
 
-    public static void initialize(RubyContext context, DynamicObject regexp, Node currentNode, ByteList setSource, int options) {
+    public static void initialize(RubyContext context, DynamicObject regexp, Node currentNode, Rope setSource, int options) {
         assert RubyGuards.isRubyRegexp(regexp);
-        setSource(regexp, setSource);
-        setOptions(regexp, RegexpOptions.fromEmbeddedOptions(options));
-        setRegex(regexp, compile(currentNode, context, setSource, Layouts.REGEXP.getOptions(regexp)));
+        final RegexpOptions regexpOptions = RegexpOptions.fromEmbeddedOptions(options);
+        final Regex regex = compile(currentNode, context, setSource, regexpOptions);
+
+        // The RegexpNodes.compile operation may modify the encoding of the source rope. This modified copy is stored
+        // in the Regex object as the "user object". Since ropes are immutable, we need to take this updated copy when
+        // constructing the final regexp.
+        setSource(regexp, (Rope) regex.getUserObject());
```

Comment:
Regexp#source does not seem perf critical, and an additional read is nothing. I think the gain comes from not having to maintain a double mapping here (and that API is likely stable and intended for this purpose).